### ISSUE TITLE
Codegen: feature request: improve reporting when user specifies too small num of pipeline stages

### ISSUE THREAD

**Issue by rdob-ant** [google/xls#1043]
TLDR: when codegen is asked to generate Verilog code with a user-provided number of pipeline stages (using `pipeline_stages` in e.g. _xls_ir_verilog()_ rule), there seems to be a lowest limit on the number of stages that must be generated, irrespective of the desired Fmax. The problem is that when the user asks it to generate code with a number of stages lower than the minimum, the codegen reports an error which is rather obscure and gives no hint of what the problem actually is. The minimal required number of stages is not reported and has to be determined experimentally via trial-and-error.

---

Example. Consider the following proc that accesses a 1R1W RAM:

```
import xls.examples.ram as ram

type RamWriteReq = ram::WriteReq;
type RamWriteResp = ram::WriteResp;
type RamReadReq = ram::ReadReq;
type RamReadResp = ram::ReadResp;

proc dpram {
    o_write_req: chan<ram::WriteReq<16, 8, 8>> out;
    i_write_resp: chan<ram::WriteResp> in;
    o_read_req: chan<ram::ReadReq<16, 8>> out;
    i_read_resp: chan<ram::ReadResp<8>> in;

    init {u16:0}

    config(
        o_write_req: chan<ram::WriteReq<16, 8, 8>> out,
        i_write_resp: chan<ram::WriteResp> in,
        o_read_req: chan<ram::ReadReq<16, 8>> out,
        i_read_resp: chan<ram::ReadResp<8>> in,
    ) {
        (
            o_write_req, i_write_resp,
            o_read_req, i_read_resp
        )
    }

    next (tok: token, state: u16) {
        let addr = state;

        let tok = send(tok, o_write_req, ram::WriteWordReq<u32:8>(addr, u8:0));
        let (tok, _) = recv(tok, i_write_resp);

        let addr = addr + u16:1;

        // Dummy read just to make IR converter happy
        let tok = send(tok, o_read_req, ram::ReadWordReq<u32:8>(u16:0));
        let (tok, _) = recv(tok, i_read_resp);

        addr
    }
}
```

Bazel rules:
```
xls_dslx_library(
    name = "dpram_dslx",
    srcs = [
        "dpram.x"
    ],
    deps = [
        "//xls/examples:ram_dslx",
    ],
)

xls_dslx_test(
    name = "dpram_dslx_test",
    dslx_test_args = {
        "compare": "none",
    },
    library = "dpram_dslx",
)

xls_dslx_ir(
    name = "dpram_ir",
    dslx_top = "dpram",
    library = "dpram_dslx",
    ir_file = "dpram_ir.ir",
)

xls_ir_opt_ir(
    name = "dpram_opt_ir",
    src = "dpram_ir.ir",
    top = "__dpram__dpram_0_next",
)

xls_ir_verilog(
    name = "dpram_verilog",
    src = "dpram_opt_ir.opt.ir",
    verilog_file = "dpram.v",
    codegen_args = {
        "module_name": "dpram",
        "delay_model": "unit",
        "pipeline_stages": "4",
        "reset": "rst",
        "use_system_verilog": "false",
        "ram_configurations": ",".join([
            "ram:1R1W:{rd_req}:{rd_resp}:{wr_req}:{wr_resp}:2".format(
                rd_req = "dpram__o_read_req",
                rd_resp = "dpram__i_read_resp",
                wr_req = "dpram__o_write_req",
                wr_resp = "dpram__i_write_resp",
            ),
        ]),
    },
)
```

Trying to build `dpram_verilog` with above configuration fails:
```
F0622 15:36:18.193552       2 codegen_main.cc:330] Check failed: ::absl::OkStatus() == (xls::RealMain(ir_path)) (OK vs. INVALID_ARGUMENT: Highest value in range fails condition of binary search.)
Target //xls/modules/test:dpram_verilog failed to build
```

The error is cryptic and does not point at the actual cause of failure. My guess is that 4 pipeline stages are not enough for this design (note that the RAM's latency is specified as `2`). Setting `pipeline_stages` to 5 or more makes codegen succeed.

**Comment by ericastor** [comment]
At the very least, we should be able to detect that the number of stages is part of the issue, automate the trial-and-error, find the minimum number of stages required, and print out the minimum number of stages required in the error message.
